src/GraphQL/Subscription/ObservableExtensions.cs

```diff
@@ -0,0 +1,253 @@
+            {
+                var queueData = new QueueData { Type = queueType, Data = task, Error = error };
+                bool attach = false;
+                lock (_queue)
+                {
+                    _queue.Enqueue(queueData);
+                    attach = _queue.Count == 1;
+                }
+
+                // start watching for an event to complete, if this is the first in the queue
+                if (attach)
+                {
+                    // for data events, start sending the data once the transformation task completes
+                    // but for error/completion events, send the event now
+                    if (task != null)
+                    {
+                        // start returning data once the first task has completed (or now if already completed)
+                        if (task.IsCompleted)
+                        {
+                            _ = ReturnDataAsync();
+                        }
+                        else
+                        {
+                            _ = task.ContinueWith(_ => ReturnDataAsync());
+                        }
+                    }
+                    else
+                        // start returning data now
+                        _ = ReturnDataAsync();
+                }
+            }
+
+            /// <summary>
+            /// Returns data from the queue in order (or raises errors or completed notifications);
+            /// executes until the queue is empty.
+            /// </summary>
+            private async Task ReturnDataAsync()
+            {
+                // grab the event at the start of the queue, but don't remove it from the queue
+                QueueData? queueData;
+                lock (_queue)
+                {
+                    // should always successfully peek from the queue here
+                    queueData = _queue.Count > 0 ? _queue.Peek() : null;
+                }
+                while (queueData != null)
+                {
+                    // process the event
+                    if (queueData.Type == QueueType.Data)
+                    {
+                        await ProcessDataAsync(queueData.Data!).ConfigureAwait(false);
+                    }
+                    else if (queueData.Type == QueueType.Error)
+                    {
+                        await ProcessErrorAsync(queueData.Error!).ConfigureAwait(false);
+                    }
+                    else if (queueData.Type == QueueType.Completion)
+                    {
+                        if (!_token.IsCancellationRequested)
+                            _observer.OnCompleted();
+                    }
+                    // once the event has been passed along, dequeue it
+                    lock (_queue)
+                    {
+                        _ = _queue.Dequeue();
+                        queueData = _queue.Count > 0 ? _queue.Peek() : null;
```

Comment:
I would switch to `TryAdd` and add internal extension method for `netstandard2.0` tfm.

Reply:
I believe you mean TryPeek. Done.  TryPeek was added to the bottom of this file as an internal method.  Lmk if you want me to move it somewhere else.

Reply:
Yes, `TryPeek`. Method is in its place.